### ISSUE TITLE
Possibility to get temporary directory from project / layout

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#15734]
Every now and then I'm in the need for creating a temporary directory outside a task or independent from a task.

One example is an `Xmllint` task that on Windows at runtime gets the Xmllint executable from a zip dependency in the company repository, unpacks it and then uses it. To avoid unpacking it 50 times and having it lying around 50 times, I want to unpack it to the temporary directory of the root project instead of the temporary directory of the task which would be different for each project where it is used.

Another example is configuring the Java Card Gradle plugin which needs a path to the Java Card Kit at configuration time already, so I have a plugin that gets the correct Kit from the repository, unpacks the zip and then configures the extension of that plugin. (lazy configuration is not really an option as the developer is not too active unless I do it myself).

Currently I do it like `val javaCardKitDirectory = rootProject.tasks.first().temporaryDir.parentFile.resolve("javaCardKit")`.
But it would be much easier and better if you could simply request the temporary directory from `project` or `project.layout` just as you can get the build directory and project directory.

**Comment by lptr** [comment]
That doesn't sound like a temporary directory but a cache directory. Why not have a dedicated unpack task, and use its output?

Another option might be to use an artifact transform to accomplish the same. That would work during configuration time as well.

**Comment by Vampire** [comment]
Dedicated unpack task is bad, as for the second described case it will not work at all and for the first described case, well, currently the `Xmllint` task is self-contained and just unpacks the zip if necessary to one location within the build. Of course it could be done by an unpack task that is added to the root project and then depended upon by using a plugin, but then the task is not self-contained anymore. That's probably not that much of a problem, just a thought. :-)

As far as the artifact transforms are concerned, thanks for that, that was exactly what I needed for the second use-case, that works amazing, I forgot about them.

But despite my concrete examples being better solved with artifact transforms (or an unpack-task), I think it could still be handy at times to have access to the path of the temporary directory from the project / layout without having a task at hand.
If you don't agree, well, close this issue, there's always a hacky way around it, at last using `java.io.tempdir` :-)

**Comment by lptr** [comment]
Alright, glad you found a good solution! I will close this issue now.

As for the more general matter, the problem is I'm not sure what Gradle is supposed to provide when it comes to temporary directories. Most of the use cases I encountered so far turned out to be like yours, and require a cache instead of a temporary directory. Another usual case is that temporary directory usage hides a `@LocalState` output, perhaps to store some incremental state (which in most cases would be better of as a real `@OutputFile`, so that it can be shared via the build cache, but that's another story).

When you need a real temporary directory, what we currently offer via `Task.getTemporaryDir()` is... problematic. We give you the same directory every time without any guarantee that it's going to be empty. Nor do we clean it up at once the task is finished with it, so it is actually not a _temporary_ directory at all, but some hidden local state that we don't clean up when loading from cache (like we do with `@LocalState`).

I'd much rather deprecate `Task.getTemporaryDir()` in favor of `@LocalState` output and `Files.createTempDirectory()`.
